crates/aptos-crypto/src/p256_ecdsa/p256_sigs.rs

```diff
@@ -0,0 +1,188 @@
+        if P256Signature::check_s_malleability(&self.to_bytes()[..]).is_ok() {
+            return self.clone()
+        };
+        let s = self.0.s();
+        let r = self.0.r();
+        // NonZeroScalar::try_from throws an error on being passed the curve order, so we use the
+        // order minus one, then add one later to compute (n-1) - s + 1 = n-s
+        //let order_minus_one = NonZeroScalar::try_from(&ORDER_MINUS_ONE[..]).unwrap();
+        //let one = NonZeroScalar::from_uint(<NistP256 as Curve>::Uint::ONE).unwrap();
+        // Dereferencing a NonZeroScalar makes it a Scalar, which implements subtraction
+        let new_s = -*s; //*order_minus_one - *s + *one;
+        let new_s_nonzero = NonZeroScalar::new(new_s).unwrap();
+        let new_sig = p256::ecdsa::Signature::from_scalars(&r, &new_s_nonzero).unwrap();
+        P256Signature(new_sig)
+    }
+}
+
+//////////////////////
+// Signature Traits //
+//////////////////////
+
+impl Signature for P256Signature {
+    type SigningKeyMaterial = P256PrivateKey;
+    type VerifyingKeyMaterial = P256PublicKey;
+
+    /// Verifies that the provided signature is valid for the provided message, going beyond the
+    /// [NIST SP 800-186](https://csrc.nist.gov/publications/detail/sp/800-186/final) specification, to prevent scalar malleability as done in [BIP146](https://github.com/bitcoin/bips/blob/master/bip-0146.mediawiki).
+    fn verify<T: CryptoHash + Serialize>(
+        &self,
+        message: &T,
+        public_key: &P256PublicKey,
+    ) -> Result<()> {
+        Self::verify_arbitrary_msg(self, &signing_message(message)?, public_key)
+    }
+
+    /// Checks that `self` is valid for an arbitrary &[u8] `message` using `public_key`.
+    /// Outside of this crate, this particular function should only be used for native signature
+    /// verification in Move.
+    ///
+    /// Checks for and rejects non-canonical signatures (r,s) where s > (n/2), where n is the group
+    /// order
+    fn verify_arbitrary_msg(&self, message: &[u8], public_key: &P256PublicKey) -> Result<()> {
+        P256Signature::check_s_malleability(&self.to_bytes())?;
+
+        public_key.0.verify(message, &self.0).map_err(|e| anyhow!("{}", e)).and(Ok(()))
+    }
+
+    fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes().to_vec()
+    }
+}
+
+impl Length for P256Signature {
+    fn length(&self) -> usize {
+        P256_SIGNATURE_LENGTH
+    }
+}
+
+impl ValidCryptoMaterial for P256Signature {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes().to_vec()
+    }
+}
+
+impl std::hash::Hash for P256Signature {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let encoded_signature = self.to_bytes();
+        state.write(&encoded_signature);
+    }
+}
+
+impl TryFrom<&[u8]> for P256Signature {
+    type Error = CryptoMaterialError;
+
+    fn try_from(bytes: &[u8]) -> std::result::Result<P256Signature, CryptoMaterialError> {
+        P256Signature::check_s_malleability(bytes)?;
```

Comment:
@alinush in David's PR you mentioned this has to happen in verification, not deserialization?

Reply:
Oh, just saw that's in verification too. It's still unclear to me high-s should be a deserialization failure, or verification failure, or it simply doesn't matter :)

Reply:
Oh, indeed, this is just a paranoid check & could be avoided.